common/fs/src/cache/tailed_file.rs

```diff
@@ -68,33 +81,33 @@ fn read_line_lossy<R: AsyncBufRead + ?Sized>(
 pin_project! {
     #[derive(Debug)]
     #[must_use = "streams do nothing unless polled"]
-    pub struct Lines {
+    pub struct LineBuilderLines {
         #[pin]
-        reader: Rc<RefCell<TailedFileInner>>,
-        bytes: Vec<u8>,
+        reader: Arc<Mutex<TailedFileInner>>,
         read: usize,
     }
 }
 
-impl Lines {
-    pub fn new(reader: Rc<RefCell<TailedFileInner>>) -> Self {
-        Self {
-            reader,
-            bytes: Vec::new(),
-            read: 0,
-        }
+impl LineBuilderLines {
+    pub fn new(reader: Arc<Mutex<TailedFileInner>>) -> Self {
+        Self { reader, read: 0 }
     }
 }
 
-impl Stream for Lines {
+impl Stream for LineBuilderLines {
```

Comment:
As I mentioned in https://github.com/logdna/logdna-agent-v2/pull/102 it would be nice if we could find a solution, like `stream::unfold()` w/ `read_until()` instead of having to code the stream internals ourselves.

These are sensitive lines of code that could hide hard to track issues.

Reply:
Yeh, I'd like to get away from the raw trait implementations. I had a look at trying to use unfold for the lower level read_until function, but I couldn't work out a way to get data out without allocating further.

The imperative, manual state machine makes it a lot easier to deal with passing pointers across polls (you just stash them in the struct)